### PR TITLE
RAD-Gen Version 0.1.0

### PR DESCRIPTION
**Pull request by StephenMoreOSU**
### What is this PR?
This pull request is to integrate the following improvements into RAD-Gen. Organized by their affected subtool. 

#### General Improvements
* Improved initialization: 
	* Cleaned up code and modified `src/common/utils.py::init_dataclass` to work for hierarchically defined inputs and recursive data structures
	* Added methodology / structure for user input initialization + modification of data sturctures to add more user parameter options
	* Moved common user input required across subtools to `Common` data structure in `src/common/data_structs.py::Common`
	* Dump out tool "state" via `.json` on invocation of `rad_gen.py` to provide way to back track outputs to input parameters.
* Unit tests enhanced and migrated to `pytest` framework
* Documentation improvements
  	* High quality docstrings for (most) functions 
  	* Better examples

#### COFFE
* Fixed bugs in tile layout wire length from BLE output to SB input
* Added support for multiple wire types FPGA architectures and more complex switch block architectures found in modern devices like stratix_iv and stratix_10
* Routing-resource-graph (RRG) parser that extracts statistical information for fabric routing mux sizes and wire load information.
* Spice lib / netlist parser to extract spice netlist information from a file and initialize a `SpSubCkt` class for each parsed `.subckt` 
* Debugging utilities to compare COFFE runs against one another
* Plotting utilities
* Large amount of refactoring to:
	* Represent / import / export / manipulate primitive spice simulation testbench objects
		* e.g. `.MEAS` statements, voltage sources, etc
	* Represent / import / export / manipulate spice simulation testbenches in a separate struct from the `SizeableCircuit` also not using raw spice.
		* Spice subckt objects required as inputs to testbenches.
	* Represent / import / export / manipulate spice netlists to create dependancies between a `.subckt` and the `.subckt` objects instantiated within it. 
	* Modified classes to use `dataclasses` 

#### ASIC DSE
* Fixed edge cases in the generation of SRAM, adjusted the cost function
* Linked ASIC sweep parameters directly to Hammer
* Reworked initialization function
* Updated to work with updated version of PrimeTime

##### Tests ran for this PR are as follows coming from the `tests/` directory:

More information on specific tests can be found by looking at the corresponding test file or looking at examples page in docs. Each test (that requires data comparison) has a golden result which it it will use to determine if it was successful.

###### Configuration Initialization
* tests/test_alu_vlsi_sweep.py::test_alu_vlsi_sweep_gen_conf_init
* tests/test_alu_vlsi_sweep.py::test_alu_sw_pt_asic_flow_conf_init
* tests/test_alu_vlsi_sweep.py::test_alu_sw_pt_parse_conf_init
* tests/test_noc_rtl_sweep.py::test_noc_sw_pt_asic_flow_conf_init
* tests/test_noc_rtl_sweep.py::test_noc_sw_pt_parse_conf_init
* tests/test_noc_rtl_sweep.py::test_noc_rtl_sweep_conf_init
* tests/test_sram_gen.py::test_single_macro_asic_flow_conf_init
* tests/test_sram_gen.py::test_single_macro_parse_conf_init
* tests/test_sram_gen.py::test_stitched_sram_asic_flow_conf_init
* tests/test_sram_gen.py::test_stitched_sram_parse_conf_init
* tests/test_sram_gen.py::test_sram_gen_conf_init
* tests/test_stratix_iv.py::test_stratix_iv_conf_init
* tests/test_ic_3d.py::test_buffer_dse_conf_init

###### ASIC DSE
* tests/test_alu_vlsi_sweep.py::test_alu_vlsi_sweep_gen
* tests/test_alu_vlsi_sweep.py::test_alu_sw_pt_asic_flow
* tests/test_alu_vlsi_sweep.py::test_alu_sw_pt_parse
* tests/test_noc_rtl_sweep.py::test_noc_rtl_sweep
* tests/test_noc_rtl_sweep.py::test_noc_sw_pt_asic_flow
* tests/test_noc_rtl_sweep.py::test_noc_sw_pt_parse
* tests/test_sram_gen.py::test_sram_gen
* tests/test_sram_gen.py::test_single_macro_asic_flow
* tests/test_sram_gen.py::test_single_macro_parse
* tests/test_sram_gen.py::test_stitched_sram_asic_flow
* tests/test_sram_gen.py::test_stitched_sram_parse

###### COFFE
* tests/test_stratix_10.py::test_stratix_10_rrg_parse
* tests/test_stratix_iv.py::test_stratix_iv_rrg_parse
* tests/test_stratix_iv.py::test_stratix_iv_passthrough
* tests/test_stratix_iv.py::test_stratix_iv_checkpoint
* tests/test_stratix_iv.py::test_stratix_iv
* tests/test_stratix_iv.py::test_stratix_iv_bram_passthrough
* tests/test_stratix_iv.py::test_stratix_iv_bram

###### IC-3D
* tests/test_ic_3d.py::test_buffer_dse
* tests/test_ic_3d.py::test_pdn

### CODE REVIEW

**Comment by AlexandreSinger** [comment]
Hi @StephenMoreOSU , thanks for adding me as a reviewer! I will review the code over the next week or so and slowly add comments. I have one overall comment that I was curious on that I thought I would mention.

I noticed that this PR is titled RAD-Gen Version 0.1.0; but I noticed that there was no CLI option to ask RAD-Gen what the current version it was in code (at least as far as I could tell). I think it would be a good idea to add this option so it would be easy for people to tell what version they are on.

Also, I noticed that RAD-Gen does not have releases (https://github.com/StephenMoreOSU/rad_gen/releases ). I think it would be a good idea to make a release of the previous version of Rad-Gen (Version 0.0.1 I guess?) before merging this in so: 1) people who were using Rad-Gen before know what version to reference and 2) allow you to know the state of Rad-Gen before your changes to Version 0.1.0. Just a suggestion to improve the infrastructure of the repository. And then you can make a release for Version 0.1.0 once it has been tested thoroughly (since you are running so many tests now), so researchers can use the stable releases for when they build off this work. After this PR is merged, in the months following, people may be developing on Master and may introduce unexpected bugs which have not been tested which will make it challenging for new users of Rad-Gen to build off.

**Comment by StephenMoreOSU** [comment]
@AlexandreSinger  Good point, I think user base is pretty small at this point but this is such a big PR that it makes sense to try track everything.

File: conda_env/env.yaml [hunk 1]
```diff
@@ -1,21 +1,36 @@
 name: rad-gen-env
+#-py-12
```
Comment (AlexandreSinger) [inline conda_env/env.yaml:2]: Leftover comment?

File: conda_env/env.yaml [hunk 2]
```diff
@@ -1,21 +1,36 @@
...
+  - matplotlib
+  - pandas
+  - numpy
+  - pyYAML
+  - python==3.12.3
+  - pandas
+  - pip
   # IC 3D 
-  - plotly=5.5.0
+  - plotly
```
Comment (AlexandreSinger) [inline conda_env/env.yaml:18]: This is WAY more work than needed, so feel free to ignore this; but since you removed the dependency versions, that means that your code should work for all versions of each of these tools (which is possible if you are only using a SMALL amount of the library; especially if it is a stable part of the library). But if you are using "unstable" features of a library, it may be a good idea to put minimum (or maximum) versions of libraries to prevent issues in the future. For example, in a year, someone may clone your repo and a recent pandas upgrade may cause issues.

File: conda_env/env.yaml [hunk 3]
```diff
@@ -1,21 +1,36 @@
...
-  - pyYAML=6.0
-  - python=3.9.13
-  - pandas=1.4.2
-  - pip=23.3.1
+  - matplotlib
+  - pandas
+  - numpy
+  - pyYAML
+  - python==3.12.3
+  - pandas
```
Comment (AlexandreSinger) [inline conda_env/env.yaml:15]: pandas depended twice. Delete.

File: conda_env/env.yaml (hunk 2 above)
Comment (AlexandreSinger) [reply conda_env/env.yaml:18]: Note: sometimes its a good idea to lock dependencies on specific version (if they are important) and then upgrading them only when you are certain that specific versions work. Makes upgrades easier and makes the code base a bit more stable since things only change version when you want them to.

File: rad_gen.py [hunk 4]
```diff
@@ -1,3 +1,5 @@
+from __future__ import annotations
```
Comment (AlexandreSinger) [inline rad_gen.py:1]: Do you still need this line, since I think you set the Python version to a high enough version? I could be wrong though.

File: rad_gen.py [hunk 5]
```diff
@@ -55,63 +58,112 @@
...
 
+# ██╗      ██████╗  ██████╗  ██████╗ ██╗███╗   ██╗ ██████╗ 
+# ██║     ██╔═══██╗██╔════╝ ██╔════╝ ██║████╗  ██║██╔════╝ 
+# ██║     ██║   ██║██║  ███╗██║  ███╗██║██╔██╗ ██║██║  ███╗
+# ██║     ██║   ██║██║   ██║██║   ██║██║██║╚██╗██║██║   ██║
+# ███████╗╚██████╔╝╚██████╔╝╚██████╔╝██║██║ ╚████║╚██████╔╝
+# ╚══════╝ ╚═════╝  ╚═════╝  ╚═════╝ ╚═╝╚═╝  ╚═══╝ ╚═════╝ 
+
+
+def init_logger():
```
Comment (AlexandreSinger) [inline rad_gen.py:76]: I personally come from the philosophy of putting the bare minimum into the "main file" of a program. Basically turning it into:
```
main:
  - parse args
  - res <- run_program
  - parse res messages
  - gracefully return
```
Perhaps this logging code is better suited for utils?

File: rad_gen.py [hunk 6]
```diff
@@ -55,63 +58,112 @@
...
 # ╚═╝  ╚═╝╚═╝  ╚═╝╚═════╝      ╚═════╝ ╚══════╝╚═╝  ╚═══╝    ╚══════╝╚═╝  ╚═╝╚══════╝ ╚═════╝    ╚═╝     ╚═╝ ╚═════╝ ╚═════╝ ╚══════╝╚══════╝
 
-def main(args: Optional[List[str]] = None) -> None:
+def main(args: argparse.Namespace | None = None) -> Tuple[ Any | None, rg_ds.Tree]:
     global cur_env
     global rad_gen_log_fd
     global log_verbosity
 
+    # Get logger into main
+    # logger = init_logger()
```
Comment (AlexandreSinger) [inline rad_gen.py:106]: Is this intended to be commented out? Perhaps it would be a good CLI option to turn this on / off?

File: src/common/utils.py [hunk 7]
```diff
@@ -1,21 +1,33 @@
-
-from typing import List, Dict, Tuple, Set, Union, Any, Type, Callable
+from __future__ import annotations
```
Comment (AlexandreSinger) [inline src/common/utils.py:1]: This file is very large and contains lots of disjoint utilities. It may be a good idea to separate them out. Tons of work for very little payoff, I know, but it would improve the readability of the codebase and make it easier to develope on in the future.

**Review by AlexandreSinger — COMMENTED** [review]
Hi @StephenMoreOSU , I have left some high-level comments. I only got up to the main file and I left a comment on Utils. These are all recommendations, a lot of it is more work than necessary, but I thought I may as well mention it; feel free to ignore.

I will continue to chip away at reviewing this PR over the coming week (I find this fun to review), but I thought I would release some of my comments early.

File: docs/Makefile [hunk 8]
```diff
@@ -3,6 +3,7 @@
 
 # You can set these variables from the command line.
 SPHINXOPTS    =
+SPHINX_APIGEN = sphinx-apidoc
```
Comment (robluo) [inline docs/Makefile:6]: When I tried to build the doc by `make html`, it complains about missing sphinxcontrib-mermaid. That can be added to the requirements.txt

Comment (vaughnbetz) [reply docs/Makefile:6]: Stephen says this is fixed ... marking resolved.

File: src/coffe/ble.py [hunk 9]
```diff
@@ -0,0 +1,1513 @@
...
+            targ_nodes = targ_nodes,
+            low_v_node = self.general_out_node,
+        )
+
+
+
+
+@dataclass
+class BLE(c_ds.CompoundCircuit):
+    name: str = "ble"
```
Comment (vaughnbetz) [inline src/coffe/ble.py:1274]: Suggest adding docstring.
Also good practice to comment each data member of a class, but don't worry about it in existing code  (aspirational goal!).

File: src/common/utils.py [hunk 10]
```diff
@@ -1,21 +1,33 @@
...
+from typing import List, Dict, Tuple, Set, Union, Any, Type, Callable, cast
```
Comment (vaughnbetz) [inline src/common/utils.py:2]: Suggest a docstring at the top of the file outlining its use and what code should go here.

**Review by vaughnbetz — COMMENTED** [review]
Code structure looks good. As time permits, adding file scope docstrings, missing docstrings on functions and return types would be useful. But I think this should be merged now.

**Comment by vaughnbetz** [comment]
Please resolve the conflicts and merge.  The suggestions for code documentation updates etc. can be handled in another PR when time permits.
